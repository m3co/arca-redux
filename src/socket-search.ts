
import Socket from 'socket.io-client';
import { v4 as uuid4 } from 'uuid';

import { State, Fields } from './types';

export interface Params {[Key: string]: string | number | null}

interface Response {
  ID: string;
  Method: "Search";
  Context: {
    Field: keyof Fields;
    Source: keyof State["Source"];
  };
  Result: Params[];
  Error: {} | null;
};

export class ARCASearchSocket {
  private readonly io: SocketIOClient.Socket;
  private ids: {[ID: string]: {
    resolve: (value: Response) => void;
    reject: (reason: string) => void;
  };} = {};

  private handleResponse = (response: Response): void => {
    const handle = this.ids[response.ID];
    if (handle) {
      if (response.Error) {
        this.ids[response.ID].reject(response.Error.toString()); // TODO
      } else {
        this.ids[response.ID].resolve(response);
      }
      delete this.ids[response.ID];
    } else {
      console.error('response', response, 'not found');
    }
  }

  public constructor(io = Socket()) {
    this.io = io;
    this.io.on('jsonrpc', this.handleResponse);
  }

  public Search = (Source: keyof State["Source"],
    Field: keyof Fields, Params: Params, ID?: string): Promise<Response> => {
    return new Promise((
      resolve: (value: Response) => void,
      reject: (reason: string) => void): void => {
      const request = {
        ID: ID || uuid4(),
        Method: 'Search',
        Context: {
          Source,
          Field
        },
        Params,
      };
      this.io.emit('jsonrpc', request);
      this.ids[request.ID] = { resolve, reject };
    })
  }
}
