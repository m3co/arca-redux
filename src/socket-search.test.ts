import Socket from 'socket.io-client';
import { ARCASearchSocket } from './socket-search';


const URL = 'http://localhost:8086';
const source = 'FACAD-BuiltInCategories';

test(`Perfoem Search of ${source}`, (done): void => {
  const io = Socket(URL);
  const socket = new ARCASearchSocket(io);

  socket.Search(source, "BuiltInCategory",
    {BuiltInCategory: "lineload"},
    '5832c751-a6a9-4cd2-9235-157d90cb95d3')
    .then(response => {
      expect(response).toMatchSnapshot();
      done();
    });

});
