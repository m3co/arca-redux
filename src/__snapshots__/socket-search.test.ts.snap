// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Perfoem Search of FACAD-BuiltInCategories 1`] = `
Object {
  "Context": Object {
    "Field": "BuiltInCategory",
    "Source": "FACAD-BuiltInCategories",
  },
  "Error": null,
  "ID": "5832c751-a6a9-4cd2-9235-157d90cb95d3",
  "Method": "Search",
  "Result": Array [
    Object {
      "BuiltInCategory": "OST_InternalLineLoads",
    },
    Object {
      "BuiltInCategory": "OST_InternalLineLoadTags",
    },
    Object {
      "BuiltInCategory": "OST_LineLoads",
    },
    Object {
      "BuiltInCategory": "OST_LineLoadTags",
    },
  ],
}
`;
